daemon/algod/api/server/lib/middlewares/auth.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package middlewares
+
+import (
+	"crypto/subtle"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/labstack/echo/v4"
+)
+
+const urlAuthFormatter = "/urlAuth/%s"
+
+// AuthMiddleware provides some data to the handler.
+type AuthMiddleware struct {
+	// Header is the token header which needs to be provided. For example 'X-Algod-API-Token'.
+	header string
+
+	// Tokens is the set of tokens which can be set to allow access.
+	tokens [][]byte
+}
+
+// MakeAuth constructs the auth middleware function
+func MakeAuth(header string, tokens []string) echo.MiddlewareFunc {
```

Comment:
since we don't really have support in multiple tokens (well, exactly two). When we would, it will need to be more sophisticated than single list so I believe this can support only one token for now.